Kubelet: Pod level Resource Management

dubstack opened the pull request:
This proposal outlines our plan for improving resource management in Kubernetes by having a Cgroup hierarchy with QoS and Pod level Cgroups. 

This is the initial proposal which broadly covers our goals and how we plan to achieve it. At this point we would really appreciate feedback from the community. 

This is tied to the upstream issue #5671. So i would request
@vishh @dchen1107 @bgrant0607  @jdef PTAL.

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

dubstack committed:
Propasal for pod level resource management

Signed-off-by: Buddha Prakash <buddhap@google.com>

vishh commented:
ok to test

vishh commented:
cc @kubernetes/sig-node

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+We have the following resource parameters for the `Bu` cgroup.
+
+'/Bu/cpu.shares = summation of all cpu requests <br>
```
nit: ' typo

derekwaynecarr replied:
@vishh @dchen1107 - when we discussed this in our F2F, we had also discussed `Bu` would be a child of `G`?  any reason for the change or am I recalling incorrectly?

derekwaynecarr replied:
NM, I remember we discussed both flat and hierarchical proposals, and this is presenting the flat.

vishh replied:
We decided to go with a flat hierarchy at the end.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+#### Top level Cgroups for QoS tiers
+
+Three top level cgroups for Guaranteed, Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. We use libcontainers cgroups manager interface to create/destroy the same. These cgroups are only created once on Kubelet initialization as a part of node setup. The Guranteed pods Cgroup would be nested directly under the root by default.
+But we will have a enable-guranteed-cgroup flag that would allow for having a specific Guaranteed Cgroup under the root.
```
typo guaranteed

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+- When a new pod is brought up, its QoS class is firstly determined.
+- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
+- This interface will be pluggable. Kubelet will support both systemd and raw cgroups based __cgroup__ drivers.
```
to clarify, this is the `--cgroup-driver` flag proposed in the systemd node spec?

vishh replied:
That's correct. We should cross-reference @derekwaynecarr's systemd node spec proposal here @dubstack

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+Guaranteed pods will be placed at the `root` cgroup by default. This can be made configurable in the future.
+
+/G/cpu.shares = summation of cpu requests of all guaranteed containers
+/G/PodUID/cpu.quota = cpu request of Pod
```
Missing:

`/G/PodUID/cpu.shares` = cpu request of pod
`/G/PodUID/cpu.quota` = cpu limit of pod

dubstack replied:
Ack.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+Pod1 and Pod2 are both classified as Guaranteed and are nested under the `/` cgroup.<br>
+
+/G/cpu.shares = 130m <br>
```
`/G/Pod1/cpu.shares` should be set as well.

dubstack replied:
Ack!

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+We have the following resource parameters for the `Bu` cgroup.
+
+'/Bu/cpu.shares = summation of all cpu requests <br>
+/Bu/PodUID/cpu.quota = Pod Cpu Limit <br>
```
Need to also set shares per pod, so `/Bu/PodUID/cpu.shares` should be set to request.

dubstack replied:
Ack!

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Best Effort Cgroup ####
+
+For pods belonging to the Best effort Qos we don't specify any cpu parameters for the pod cgroup.<br>
```
I would just say you don't set any quota.

dubstack replied:
Yeah that's definitely more precise. :+1:

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+For pods belonging to the Best effort Qos we don't specify any cpu parameters for the pod cgroup.<br>
+
+/BE/cpu.shares = 2 <br>
```
For those less familiar, an explanation on why this is set to 2 would be helpful.

dubstack replied:
Ack! Would add a short note about the same.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+/BE/cpu.shares = 2 <br>
+/BE/cpu.quota= not set
+/BE/m.limit_in_bytes = Allocatable - summation of memory requests/limits of guaranteed pods <br>
+/BE/PodUID/m.limit = no limit <br>
```
What if a user specifies a pod with just a cpu request or just a memory request, and not both?

For example, provide an example of a pod that has a container that just does:

requests=cpu=100m

and another with

requests=memory=500M

derekwaynecarr replied:
re: above, I assume they are just enforced via the local container scope, but it would be good to clarify.

dubstack replied:
Correct me if I am wrong did you mean to ask: Which QoS cgroup do pods with just a cpu request or just a memory request, and not both are brought under? 
If thats the case The [Quality of Service](../../docs/design/resource-qos.md) doc covers in detail how the pod classification is done.
In that case I would add a reference to the same in the propasal.

vishh replied:
@derekwaynecarr After my recent changes to pod QoS, the example pod you mentioned will be classified as `Burstable`.

derekwaynecarr replied:
@vishh - I may have missed that, but I would have thought that a minimum request would have had to been made for both cpu and memory to be burstable.

derekwaynecarr replied:
@vishh - nm, it makes sense as you described.  so lack of cpu would end up with shares=2.  In that case, if you just had a Bu and BE pod on the node, they would get equal CPU time.  I think that would be good to doc here or elsewhere as it takes some thought to recognize that lack of CPU will get you default min shares in the Bu class.

derekwaynecarr commented:
It would be good if this proposal can give an outline of the entire system cgroup hierarchy.

For example, it should include something more akin to:
https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/kubelet-systemd.md#proposed-cgroup-hierarchy

I think we also need a description on how we would at minimum enforce CPU shares for system-reserved and kube-reserved.  Otherwise, they will get the default shares value of 1024 which may or may not reflect desired outcome.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+/BE/cpu.shares = 2 <br>
+/BE/cpu.quota= not set
+/BE/m.limit_in_bytes = Allocatable - summation of memory requests/limits of guaranteed pods <br>
```
Allocatable - summation of memory requests for G and Bu pods

derekwaynecarr replied:
So this will have an interesting side effect that we should clarify.

The kubelet is more likely to OOM kill existing BE pods as more G and Bu pods are scheduled even if the G and Bu pods are using less than their request since we will have to dynamically reduce the size of BE m.limit_in_bytes.

The existing eviction behavior will not cause those BE pods to get moved since they will not be able to hog memory.

vishh replied:
Yeah. As `G` pods are added, `BE` pods are more susceptible to being OOM Killed. This policy attempts to provide best performance for `G` pods. In the future, we can let users configure the level of over-commit.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+'/Bu/cpu.shares = summation of all cpu requests <br>
+/Bu/PodUID/cpu.quota = Pod Cpu Limit <br>
+/Bu/m.limit_in_bytes = Allocatable - summation of memory requests/limits of guaranteed pods <br>
```
Allocatable - sum memory requests G pods

derekwaynecarr replied:
At one point, we discussed setting Bu/m.soft_limit_in_bytes = sum of memory requests of Bu pods

dubstack replied:
Would you suggest having it? We felt that soft limits don't really offer much gains during memory pressure.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Top level Cgroups for QoS tiers
+
+Three top level cgroups for Guaranteed, Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. We use libcontainers cgroups manager interface to create/destroy the same. These cgroups are only created once on Kubelet initialization as a part of node setup. The Guranteed pods Cgroup would be nested directly under the root by default.
```
This is not actually supported yet in libcontainer.
(see https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/cgroups.go#L11)

Apply requires a pid , and we need something that can work in the absent of a pid, and can produce slices / etc. generically.  For systemd at least, it's entirely focused on only ever creating a scope, and not general slice management.

dubstack replied:
Yeah right! We don't have support for slice management in systemd yet. So we still have to figure out on how to add support for the same. 

But for raw cgroup fs we can pass out a fake pid (0 in this case) to `Apply()` for creating cgroups. And in general as far I have seen it has support for basic cgroup management such as destruction, update etc. Please correct me if I am wrong.

But yeah having a something which can work in the absence of a pid would be the best way to go. I will discuss offline with @vishh and get back.

vishh replied:
We ideally need an explicit API from libcontainer. The hack that @dubstack mentioned should work though.

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+## Rollout Plan
+
+This feature will be opt-in in v1.4 and an opt-out in v1.5. We recommend users to drain their nodes and opt-in, before switching to v1.5, which will result in a no-op when v1.5 kubelet is rolled out.
```
+1

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+Kubernetes currently supports container level isolation only and lets users specify resource requests/limits on the containers [Compute Resources](../../docs/design/resources.md). Kubelet creates cgroups sandbox (via docker) for all these containers.
+Each pod can belong to a separate Quality of Service class. There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole - On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical. Containerd-shim in the case of docker can be charged to the pod instead of the machine.
```
without definition, `containerd-shim` doesn't make a lot of sense here

dubstack replied:
Ack! I would add a definition for the same.

dubstack commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Container level cgroups
+
+Have docker manager create container cgroups under pod level cgroups. This can be easily implemented as now docker has support for specifying parent flag of cgroups.
```
Yeah! Thanks for pointing that out. I would add a point about that.

dubstack commented:
@derekwaynecarr Thanks a lot for your comments. Yeah I too feel that a a general cgroup hierarchy outline similar to one in the kubelet-systemd propasal would be a really good idea.

dubstack commented:
@derekwaynecarr  I have addressed your comments. PTAL. Thanks.

Also we thought that its better if we brought up the guaranteed pods directly under the `Root` cgroup by default and __not__ have a separate `G` cgroup for the guaranteed qos class (This again is in line with our goal of improving resource guarantees to the guaranteed pods). We would have a flag `guaranteed-qos-at-root` which would allow the user to explicitly specify if he needs a separate `G` cgroup for the Guaranteed Qos class. I have modified the examples to reflect this behaviour. 
Does that sound good to you? Do you have some cases in your mind where this decision might turn out to be bad?

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+
+### Proposed cgroup hierarchy:
+
+For the initial implementation we will only support limits for cpu and memory resources.<br>
```
a wild br appears. Fix this?

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+  |   ...
+  +- ...
+  |
+  +- Bu
```
what is BE and Bu below?

derekwaynecarr replied:
this was shorthand for Bustable and BestEffort, agree it should be defined earlier in the spec.

philips commented on docs/proposals/resource_management.md:
````diff
@@ -0,0 +1,387 @@
+  |
+```
+
+- We have a top level Qos cgroup for the `Bu`(Burstable) and `BE`(Best Effort) Qos classes.
````
probably good to move this above the diagram.

dubstack replied:
Ack!

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+- But we __dont__ have a separate cgroup for the `G`(Guaranteed) QoS class by default. Guaranteed pod cgroups are by default brought up directly under the `Root` cgroup. Although we have a `guaranteed-qos-at-root` flag to allow users to specify if a separate `G`(Guaranteed) cgroup is required.
+- Each pod has its own cgroup which is nested under the cgroup matching the pod's Qos class.
+- All containers brought up by the pod are nested under the pod's cgroup.
+Examples in the next section.
```
Might want a newline above for markdown renders

dubstack replied:
Ack!

philips commented on docs/proposals/resource_management.md:
````diff
@@ -0,0 +1,387 @@
+We propose the following hierarchy:
+
+```
+$ROOT
````
Root isn't really defined.

vishh replied:
I believe `$ROOT` here refers to `/`. @dubstack can you make it explicit?

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+/PodUID/m.limit = memory limit of Pod <br>
+
+Note: In case guaranteed-qos-at-root flag is used to specify a separate Guaranteed cgroup, then apart from the above parameters we additionally have:<br>
+/G/cpu.shares = summation of cpu requests of all guaranteed containers <br>
```
more br's above

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+
+/PodUID/cpu.quota = cpu limit of Pod
+/PodUID/cpu.shares = cpu request of Pod
+/m.limit_in_bytes = summation of memory limits of all guaranteed containers
```
I don't quite understand this hierarchy and how it is defined.

dubstack replied:
Is the whole cgroup hierarchy in general is not very clearly conveyed or its just the idea behind the cgroup resource parameters that we have for the various cgroups? or both? Can you point at specific parts of the hierarchy that is not clear?
I guess I will have to redefine various parts of the hierarchy information to make it more clear. Your pointers would be really helpful.  Thanks!

vishh replied:
This is incorrect. We don't have to set memory limits for `G` QoS Class. @dubstack can you remove this setting?

dubstack replied:
Re: above, that particular line was added by mistake. Sorry about that. Can you go through the proposed hierarchy once again and see if it makes sense now?

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+/BE/cpu.shares = 2 <br>
+/BE/cpu.quota= not set <br>
+/BE/m.limit_in_bytes = Allocatable - 7Gi <br>
+/BE/Pod5/m.limit = no limit <br>
```
more brs

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+/Bu/Pod4/cpu.shares = 10m <br>
+/Bu/m.limit_in_bytes = Allocatable - 5Gi <br>
+/Bu/Pod3/m.limit = 3Gi <br>
+/Bu/Pod4/m.limit = 2Gi <br>
```
more brs

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+
+#### Rkt runtime
+
+We want to have rkt create pods under a root QoS class that kubelet specifies, and set pod level cgroup parameters mentioned in this proposal by itself.
```
rkt simply relies on systemd to setup cgroups. We can easily modify rkt to join a pre-setup cgroup.

philips replied:
Oops, rkt executes in whatever cgroup it is started in. So, the kubelet + rkt integration just needs to know the cgroup and it will join it. This is easy.

dubstack replied:
Yeah providing the parent cgroup should be easy. Sounds good.

philips commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,387 @@
+
+#### Container level cgroups
+
+Have docker manager create container cgroups under pod level cgroups. For determining the parent pod cgroup name and path we have a separate implementation for systemd and raw cgroup systems. This can be easily implemented as now docker has support for specifying parent flag of cgroups. We don't have support for slice management in systemd yet. So we will have to add support for parent flag and g ...
```
I am confused by "So we will have to add support for parent flag and general slice management in systemd."

Do we main to say "Kubelet will need to add support for the parent flag and systemd slice management to get this working."?

dubstack replied:
Yeah. Would go ahead and rephrase it.

dubstack commented:
Hey Brandon, thanks for your comments. I have made the suggested changes. 
Can you briefly go through the proposal once again to see if the cgroup hierarchy description makes sense now. 

Also I would really appreciate any other suggestions that you might have. Thank you.

ronnielai commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+            resources:
+                limits:
+                    cpu: 100m
+                        memory: 1Gi
```
Spacing for memory is off.

dubstack replied:
Ack! Thanks.

ronnielai commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+
+Pod3 and Pod4 are both classified as Burstable and are hence nested under the Bu cgroup  
+
+/Bu/cpu.shares = 130m  
```
30m or 130m?

dubstack replied:
Yeah it would be 30m. Thanks!

ronnielai commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+
+/BE/cpu.shares = 2  
+/BE/cpu.quota= not set  
+/BE/m.limit_in_bytes = Allocatable - 7Gi  
```
Allocatable - 10Gi? (5 from Guaranteed and 5 from Burstable)

dubstack replied:
Oh sorry I wrongly mentioned:  
/BE/m.limit_in_bytes = Allocatable - summation of memory "limits" of Guaranteed and Burstable pods
It would actually be Allocatable - summation of memory "requests" of Guaranteed and Burstable pods. 
So it would Allocatable -7Gi (5Gi request from Guaranteed and 2Gi from Burstable) 

Just to explain why we have summation of requests: Some Burstable containers might not have a limit specified for them in the Pod Spec, in that case their limit is just set to Node resource capacity. So having Allocatable - limits really would be incorrect.

ronnielai replied:
Say pod4 has no memory limit set, should its m.limit be set to node resource capacity or (node resource capacity - summation of memory limit/request of guaranteed pods)?

dubstack replied:
In this case setting the limits as  `node resource capacity - summation of memory limit/request of guaranteed pods)`
is not required as we are already setting limits on the Burstable Qos class cgroup as `node resource capacity - summation of memory limit/request of guaranteed pods)` 
So in case of memory contention the pods inside the Burstable Qos cgroup would be limited to 
`node resource capacity - summation of memory limit/request of guaranteed pods)` by the parent Burstable cgroup

derekwaynecarr commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+
+Examples in the next section.  
+
+#### Quality of Service classes
```
move this section above proposed cgroup hierarchy

dubstack replied:
SG.

derekwaynecarr commented:
Can you describe how `--cgroup-root` is expected to work here?  Is it deprecated in your view?  Is `--cgroup-root` basically `/` in this proposal so it will have a different default?

vishh commented:
@derekwaynecarr We can keep `--cgroup-root` around. All of the QoS cgroups will be underneath `--cgroup-root`. I'm not sure how `system-reserved` and `kube-reserved` should be handled in this situation though. Since we expect users to specify those cgroups, we can let them manage the interactions between `--cgroup-root` and `system-reserved & `kube-reserved`. WDYT?

asridharan commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+As of now [Quality of Service](../../docs/design/resource-qos.md) is not enforced at a pod level. Excepting pod evictions, all the other QoS features are not applicable at the pod level.
+To better support QoS, there is a need to add support for pod level resource accounting in Kubernetes.
+
+We propose to have to have a unified cgroup hierarchy with pod level cgroups for better resource management. We will have a cgroups hierarchy with top level cgroups for the three QoS classes.
```
s/to have to have/to have

dubstack replied:
Thanks!

ConnorDoyle commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+Each pod can belong to a separate Quality of Service class. There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole - On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation that allows it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
```
s/repartented/reparented

dubstack replied:
Thanks!

ConnorDoyle commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+
+#### Pod level Cgroup creation and deletion (Docker runtime)
+
+- When a new pod is brought up, its QoS class is firstly determined.
```
Re: automatic QoS classification. Making the API backwards-compatible is a great goal, and auto classification seems like a good way to go. However, as a user submitting critical pods it would be nice to guarantee the right isolation strategy is going to be applied without replicating the classification logic on the client side. What do you think about adding an optional explicit QoS class field to the pod spec? When absent, it could be filled in on the user's behalf. When present it could be used to validate the resource shape conforms to the intended class.

dubstack replied:
Yeah I agree. We would like to guarantee the right isolation strategy to all pods in all scenarios.
And adding an explicit QoS class field to the pod Spec is surely one simple way to achieve that. 
But there are a few problems with that approach that come to my mind. 
In one scenario if a user schedules a pod with Guaranteed Qos class with no memory limits specified. And if this pod starts hogging a lot of memory this would affect other pods in the Guaranteed Qos class.

According to the current policy if a user wants to schedule a critical pod, he would be required to specify the limits for all the containers in the Pod Spec. In that case the pod would deterministically  be classified as a  Guaranteed pod.

dubstack replied:
Re: I realize the benefits of having this. You suggest to have this flag as a safety check and it would alert the user if the resource shape doesn't conform to the intended class. 
In that case it should be a good addition. But we would not want to have this in the scope of the current proposal. I believe we might want to have it in [Resource-Qos](https://github.com/kubernetes/kubernetes/blob/master/docs/design/resource-qos.md)
cc @vishh

ConnorDoyle replied:
> What should be done in case the resource shape doesn't conform to the intended class.

Bailing out as early as possible in all cases sounds reasonable. For example the API server could respond with status code 400 and explain why the pod is invalid.

It would add complexity but could tip the feature more toward "hard to use wrong". Just a thought. Maybe there are alternatives to consider.

dubstack replied:
Yeah that sounds reasonable. Thanks a lot for raising the point.

ronnielai replied:
I think adding another QOS class field to validate the spec may be too confusing. It means that the user has to know the subtle relationship between the QOS class field and resource limit/request to create a valid spec.

dubstack replied:
>I think adding another QOS class field to validate the spec may be too confusing.

If the user is not really concerned about the QOS class thats assigned to his pod, he might just skip it. So it wont be confusing for these users.

> It means that the user has to know the subtle relationship between the QOS class field and resource limit/request to create a valid spec.
In case a user wants to schedule a critical pod, to which he would absolutely want to guarantee the specified resources. He would want to use the flag even if its at the cost of understanding the logic of pod QoS classification. (Which is pretty simple in that he just has to specify limits for all the containers in the pod)

ConnorDoyle replied:
Another option could be to default pods to burstable unless they explicitly elect to be guaranteed or best-effort. That is also backward-compatible and has no subtle resource rules. It would also decrease coupling between the semantics of a QoS class and the implementation; one could imagine a future where there are better controls than CFS shares and quota. In fact, it might be worthwhile to consider decoupling the isolation here and making interpretation of the classes pluggable. The strategy from this proposal could be the first and the default, but it would leave the door open to custom behavior.

vishh replied:
@ConnorDoyle I'd suggest opening a separate issue for changing QoS semantics. 

The QoS classes are currently inferred from `request`` and `limits` set on the pod. How QoS is enforced can change over time. This proposal is based on existing upstream kernel isolation primitives. If new primitives show up in the future, we can definitely incorporate them.

If you don't like the way QoS is currently inferred, kindly open a separate issue and that way I can involve the rest of the community who helped design this feature.

As for making things pluggable, the strategy we have been following as of now is to have a stable upstream implementation first, before making plugins. If we start with plugins right away, we are concerned about not being able to make progress. To make it easier to move to plugins in the future, it will be helpful to ensure that the implementation of this proposal is modular enough to make the QoS policy plugin friendly in the future.

vishh replied:
@ConnorDoyle If you do open a separate issue, it will be very helpful if you can explain how changes to the QoS policy will help end users.

ConnorDoyle replied:
Hey @vishh, your points are totally fair. Didn't mean to disrupt progress with those suggestions.

derekwaynecarr replied:
@ConnorDoyle this would make server side defaulting difficult in many cases.  I am not sure I agree with passing a QoS field in pod spec.

timothysc replied:
There is a piece of the puzzle here that's missing imo.  Many cluster managers have policy defined via submission expressions, to fill in gaps that may exist, and allow for explicit specification (pod spec).  In the case of kubernetes, it would likely be some form of admission control plug.  This enables operators the ability to control policy in a way that makes sense to them.  

Why?  people are going to game the system...

vishh replied:
We have limit ranges which are admissions control plugins for compute
resources.

On Fri, Jun 10, 2016 at 1:17 PM, Timothy St. Clair <notifications@github.com
> wrote:

> In docs/proposals/resource_management.md
> <https://github.com/kubernetes/kubernetes/pull/26751#discussion_r66674618>
> :
>
> > +## Implementation Plan
> > +
> > +The implementation plan is outlined in the next sections.
> > +We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default atleast in v1.4.
> > +
> > +#### Top level Cgroups for QoS tiers
> > +
> > +Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't have equivalent support for slice management in libcontainer yet. So we will be adding support for the same in the Kubelet. These cgroups are only created once on Kubelet initialization as a part of node setup.
> > +
> > +`Note:
> > +The Guaranteed pods cgroup would be nested directly under the root by default.
> > +But we will have a "enable-guaranteed-cgroup" flag that would allow for having a specific Guaranteed Cgroup under the root.`
> > +
> > +#### Pod level Cgroup creation and deletion (Docker runtime)
> > +
> > +- When a new pod is brought up, its QoS class is firstly determined.
>
> There is a piece of the puzzle here that's missing imo. Many cluster
> managers have policy defined via submission expressions, to fill in gaps
> that may exist, and allow for explicit specification (pod spec). In the
> case of kubernetes, it would likely be some form of admission control plug.
> This enables operators the ability to control policy in a way that makes
> sense to them.
>
> Why? people are going to game the system...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26751/files/204a2d0f06873eaaccf19ed953338d968c3c9390#r66674618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKEMNjWD2iY1iKHCAlMQvKYs-51EWks5qKcY9gaJpZM4ItEyP>
> .
>

ConnorDoyle commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+
+#### Guaranteed
+
+Guaranteed pods will be placed at the `Root` cgroup by default. We would have a flag `guaranteed-qos-at-root` which would allow the user to explicitly specify if he needs a separate `G` cgroup for the Guaranteed Qos class. So by default the flag is set to true.
```
A note about the rationale behind handling the guaranteed class differently (either here or in the implementation section) would be helpful. Maybe it's mentioned but if so I didn't catch it.

dubstack replied:
Its not explicitly mentioned anywhere. Adding a note would be a good idea. Thanks

ConnorDoyle commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,381 @@
+ - Existing Quality of Service support does not apply to pods as a whole - On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation that allows it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
+ - If a container exits, all its anonymous pages (tmpfs) gets accounted to the machine (root). With pod level cgroups, that usage can be attributed to the pod.
```
s/gets/get

ConnorDoyle commented on docs/proposals/resource_management.md:
````diff
@@ -0,0 +1,395 @@
+
+```
+/Bu/cpu.shares = summation of cpu requests of all Bu pods  
+/Bu/PodUID/cpu.quota = Pod Cpu Limit  
````
Seems like to maximize utilization, all burstable pods should have the same quota. This setting should not impact the guaranteed class' access to the CPU. If that is true, would it make sense to set only aggregate quota on the /Bu cgroup instead? Or, are there other goals of the burstable limit besides protecting the guaranteed class?

ConnorDoyle commented on docs/proposals/resource_management.md:
```diff
@@ -0,0 +1,395 @@
+## Implementation Plan
+
+The implementation plan is outlined in the next sections.
+We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default atleast in v1.4.
```
s/atleast/at least

vishh commented:
This proposal LGTM. We can continue iterating on it. @dubstack I suggest starting with the implementation right away.

derekwaynecarr commented:
@vishh - i think `cgroup-root` should be redefined as root of qos hiearchy and not actually pods.  If specified, `kubelet` could set `Allocatable` values to that location.

dubstack commented:
@derekwaynecarr I spoke to @vishh offline and he agrees to your suggestion. I will include this in my work items for this proposal. Add a note to the proposal about the same.

derekwaynecarr commented:
I want to clarify how I intend behavior to work on `systemd` platform that uses `systemd` to manage the `cgroupfs` because it will be lazy in nature.

If I run the kubelet as follows:

```
--cgroup-driver=systemd
--cgroup-root=machine.slice
```

Bootstrapping:

`kubelet` invokes a `QosContainerManager` and it has an implementation for `systemd` or `cgroupfs`. I am using systemd, I would get the following *transient slices* created: `burstable.slice`, `besteffort.slice`, and they would be parented under the `machine.slice`.

If you look at the cgroupfs tree, you will not see an actual representation of them yet.  For example, `/sys/fs/cgroup/memory/machine.slice/machine-burstable.slice` would not actually exist yet.  This is because `systemd` keeps the cgroup tree clean when there are no active processes in it.

Pod Lifecycle:

When a pod is created, I will create a slice as a *transient unit* for the pod under the appropriate qos.  I suspect in code, I will interact with `PodContainerManager` that has an implementation for `systemd`.  It will interface with the `QosContainerManager` to find the name of the container for the pod qos class, and it will create a slice `pod<uuid>.slice` whose parent is the appropriate qos class.

If you look at the cgroupfs tree, you will not see an actual representation of them yet.  For example, `sys/fs/cgroup/memory/machine.slice/machine-burstable.slice/machine-burstable-pod<uuid.slice` will not materialize yet, but systemd will be aware of the units.

Container Lifecycle:

When a container is created, and the parent container is `machine-burstable-pod<uuid>.slice`, the container manager will create a `scope` whose parent is that slice, and launch the processes associated with that container in that slice.

At that time, will you finally see the representation on the `/sys/fs/cgroup` tree.

`sys/fs/cgroup/memory/machine.slice/machine-burstable.slice/machine-burstable-pod<uuid.slice/<runtime>containerXyZ.scope/`

Hope this helps as I think it impacts the internal implementation to not assume a particular manifestation.

derekwaynecarr commented:
I guess I will temper my previous statement some, we could choose to activate each slice at point of creation which should result in the manifestation in the sys tree.  I guess we can work that out in the implementation PRs.

vishh commented:
@dubstack Can we get this PR completed and merged?

dubstack commented:
Yes as discussed offline I will remove support for the guaranteedQosAtRoot flag. By default all guaranteed pod containers would lie directly under the cgroup-root. If a user wants to have the guaranteed pod containers in a specific path, the cgroup root can be configured accordingly.
And I will also make sure that all nits are addressed. Then we can get this merged.

derekwaynecarr commented:
Poke when updates are done

On Wednesday, June 22, 2016, Buddha Prakash <notifications@github.com>
wrote:

> Yes as discussed offline I will remove support for the guaranteedQosAtRoot
> flag. By default all guaranteed pod containers would lie directly under the
> cgroup-root. If a user wants to have the guaranteed pod containers in a
> specific path, the cgroup root can be configured accordingly.
> And I will also make sure that all nits are addressed. Then we can get
> this merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26751#issuecomment-227908465>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF8dbCt-6Nrs6QtkpPM-_gJhHRahfnMIks5qOcaSgaJpZM4ItEyP>
> .
>

dubstack commented:
@derekwaynecarr absolutely.

dubstack commented:
@derekwaynecarr @vishh I have addressed all comments. PTAL.

I will squash commits after addressing your final nits.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+As of now [Quality of Service](../../docs/design/resource-qos.md) is not enforced at a pod level. Excepting pod evictions, all the other QoS features are not applicable at the pod level.
+To better support QoS, there is a need to add support for pod level resource accounting in Kubernetes.
+
+We propose to have to have a unified cgroup hierarchy with pod level cgroups for better resource management. We will have a cgroups hierarchy with top level cgroups for the three Quality of Service(QoS) classes Guaranteed, Burstable and BestEffort. Pods (and their containers) belonging to a QoS class will be grouped under these top level QoS cgroups. And all containers in a pod are nested under t ...
```
"We propose to have a unified"

derekwaynecarr replied:
s/cgroups hierarchy/cgroup hierarchy

derekwaynecarr replied:
s/pod cgroups./pod cgroup.

derekwaynecarr replied:
s/for the three Quality of Service/for the three QoS

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+## Motivations
+
+Kubernetes currently supports container level isolation only and lets users specify resource requests/limits on the containers [Compute Resources](../../docs/design/resources.md). Kubelet creates cgroups sandbox (via docker) for all these containers.
```
The `kubelet` creates a cgroup sandbox (via it's container runtime) for each container.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+## Introduction
+
+As of now [Quality of Service](../../docs/design/resource-qos.md) is not enforced at a pod level. Excepting pod evictions, all the other QoS features are not applicable at the pod level.
```
As of now Quality of Service (QoS) is not enforced at a pod level.

(define the acronym at first use, and then use it throughout this document consistently)

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+
+There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole. On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
```
/s/Quality of Service/QoS

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole. On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
```
The overhead of setting up and managing the pod sandbox.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole. On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation allowing it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
```
The containerd-shim is only present in docker version > 1.11.

In addition, `docker 1.11` associates the `containerd-shim` with the cgroup that contains the `docker-engine` and not the cgroup for the end-user container.  This is a problem I am not sure `docker` will fix, but is something that we should avoid with other run-time choices.  Absent a change in the `docker` runtime, this shim is overhead to the `kube-reserved` or `system-reserved` node allocation.  I find this text confusing absent a plan to say how the containerd-shim is reparented (which i think should not be fixed in kubelet, and fixed in docker)

nit: typo /repartented/reparented

dubstack replied:
What course of action would you suggest. If its not fixed in docker, we would like to fix this in the kubelet. 
cc @vishh @dchen1107

euank replied:
I think the best first course of action is to try for it in docker. If we can get a `--containerd-cgroup-parent` flag or just change the default then we can fix this better and more easily than in the kubelet. At a glance I didn't see such an issue on their side, so perhaps we should file one.

dubstack replied:
I will go ahead and file an issue with docker

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation allowing it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
+ - If a container exits, all its anonymous pages (tmpfs) gets accounted to the machine (root). With pod level cgroups, that usage can also be attributed to the pod.
+ - Let containers share resources - with pod level limits, containers can choose to burst on demand and still be capped at their pod limits.
```
This is better understood with an example.

Let containers share resources - with pod level limits, a pod with a Burstable container and a BestEffort container is classified as Burstable.  The BestEffort container is able to consume slack resources not used by the Burstable container, but still be capped by the overall pod level limits.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+A pod can belong to one of the following 3 QoS classes: Guaranteed, Burstable, and Best-Effort, in decreasing order of priority.
+We use the following denotation:
+
+`G⇒ Guaranteed QoS, Bu⇒ Burstable QoS, BE⇒ Besteffort Qos`
```
s/Besteffort Qos/BestEffort QoS

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+#### Guaranteed
+
+Guaranteed pods will be placed at the `$Root` cgroup by default. `$Root` is the system root ie. "/" by default and if `--cgroup-root` flag is used then we use the specified cgroup-root as the `$Root`.
+`Note: The cgroup-root flag would allow the user to configure the root of the QoS cgroup hierarchy. Hence cgroup-root would be redefined as the root of QoS cgroup hiearchy and not actually pods.`
```
add a new line, get rid of the backticks, and maybe just bold the 

/s/hierarchy/hiearchy

s/and not actually pods/and not actually containers

(we did not have pod level cgroups before, the kubelet text was confusing)

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
````diff
@@ -0,0 +1,394 @@
+`Note: The cgroup-root flag would allow the user to configure the root of the QoS cgroup hierarchy. Hence cgroup-root would be redefined as the root of QoS cgroup hiearchy and not actually pods.`
+
+```
+/PodUID/cpu.quota = cpu limit of Pod  
````
This is not clarified, but does this mean if I have a pod with UID: `5f9b19c9-3a30-11e6-8eea-28d2444e470d`

which cgroup path do I get:

1. /5f9b19c9-3a30-11e6-8eea-28d2444e470d/cpu.quota
2. /pod5f9b19c9-3a30-11e6-8eea-28d2444e470d/cpu.quota
3. /pod-5f9b19c93a3011e6-8eea28d2444e470d/cpu.quota

I am hoping 3 so I can inspect /sys/fs/cgroup and know this is a pod...

derekwaynecarr replied:
This is more a question for how the implementation will allow me to clearly see the cgroupfs that this cgroup aligns to a pod, etc. and not just getting opaque ids.

derekwaynecarr replied:
If we can just state what PodUID expands to in one spot, that would be good, and we can then use the shorthand throughout.  same with the qos level cgroup.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
````diff
@@ -0,0 +1,394 @@
+We have the following resource parameters for the `Bu` cgroup.
+
+```
+/Bu/cpu.shares = summation of cpu requests of all Bu pods  
````
I know this is explanatory, but on a systemd system, I would like to be able to type:

`systemctl status burstable.slice`

and get aggregate usage of that tier, so just a long winded way of saying I would not want to abbrev this in implementation.

dubstack replied:
Really sorry but I am not sure if i understand what you mean by " I would not want to abbrev this in implementation."

derekwaynecarr replied:
I mean I do not want a slice on my machine called "Bu", so when we implement this, it should have human-readable fully qualified names like "burstable.slice"

dubstack replied:
Yes thats exactly how I planned to have it.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+### Hierarchy Outline
+
+- "$Root" is the system root of the node ie. "/" by default and if `--cgroup-root` is specified then the specified cgroup-root is used as "$Root".
+- We have a top level Qos cgroup for the `Bu`(Burstable) and `BE`(Best Effort) Qos classes.
```
/s/Qos/QoS

I imagine by this point `Bu` is understood in the proposal and you don't need to redefine it (same for others)

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+- "$Root" is the system root of the node ie. "/" by default and if `--cgroup-root` is specified then the specified cgroup-root is used as "$Root".
+- We have a top level Qos cgroup for the `Bu`(Burstable) and `BE`(Best Effort) Qos classes.
+- But we __dont__ have a separate cgroup for the `G`(Guaranteed) QoS class. Guaranteed pod cgroups are brought up directly under the `Root` cgroup.
+- Each pod has its own cgroup which is nested under the cgroup matching the pod's Qos class.
```
/s/QoS

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+  +- System-reserved
+  |   |
+  |   +- system
+  |   +- ...
```
docker here is valid too.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+QoS Policy Design Decisions
+
+- This hierarchy highly prioritizes resource guarantees to the Guaranteed over Burstable and Best Effort Pods.
```
s/Pods/pods

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+QoS Policy Design Decisions
+
+- This hierarchy highly prioritizes resource guarantees to the Guaranteed over Burstable and Best Effort Pods.
+- By not having a separate cgroup for the Guaranteed class, the hierarchy allows the Guaranteed pods to burst and utilize all of Node's Allocatable capacity.
```
please stop mixing G and Guaranteed, etc.  just be consistent in using the shorthand ;-)

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+## Implementation Plan
+
+The implementation plan is outlined in the next sections.
+We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default at least in v1.4.
```
I would prefer looking at all cgroup flags together in `kubelet -h` calls.

As a result, recommend `cgroup-per-qos=true|false` so its next to `cgroup-root` in help.

dubstack replied:
@vishh What would be your recommendation?

derekwaynecarr replied:
fyi: we followed a similar approach with eviction, the kubelet takes a ton of flags, anything we can do to make them appear grouped based on our naming convention, the better it is imho

dubstack replied:
Makes sense! Agree!

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+#### Top level Cgroups for QoS tiers
+
+Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't hav ...
```
nit: add space cgroup management(...)

Please clarify that on systemd these cgroups are transient units and will not survive reboot.

(this ensures kubelet cleans up after itself)

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+#### Pod level Cgroup creation and deletion (Docker runtime)
+
+- When a new pod is brought up, its QoS class is firstly determined.
+- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
```
will this happen before or after image pulling?

/cc @vishh - I would like to be able to in the future charge image pulling to the pod cgroup.

vishh replied:
This will happen before kubelet performs any activity for a pod. Ideally,
we can charge all pod related activities including image pulling, volumes
creation, network setup, etc., to the pod level cgroups.

On Fri, Jun 24, 2016 at 10:45 AM, Derek Carr <notifications@github.com>
wrote:

> In docs/proposals/pod-resource-management.md
> <https://github.com/kubernetes/kubernetes/pull/26751#discussion_r68434206>
> :
>
> > +- Also we don't set any quota on the cpu resources as the containers on the `BE` pods can use any amount of free resources on the node.
> > +- Having memory limit of `BE` cgroup as (Allocatable - summation of memory limits of Guaranteed and Burstable pods) would result in `BE` pods becoming more susceptible to being OOM killed. As more `G` and `Bu` pods are scheduled kubelet will more likely kill `BE` pods, even if the `G` and `Bu` pods are using less than their request since we will be dynamically reducing the size of `BE` m.limit_in_bytes. But this allows for better memory guarantees to the `G` and `Bu` pods.
> > +
> > +## Implementation Plan
> > +
> > +The implementation plan is outlined in the next sections.
> > +We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default at least in v1.4.
> > +
> > +#### Top level Cgroups for QoS tiers
> > +
> > +Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't have equivalent support for slice management in libcontainer yet. So we will be adding support for the same in the Kubelet. These cgroups are only created once on Kubelet initialization as a part of node setup.
> > +
> > +#### Pod level Cgroup creation and deletion (Docker runtime)
> > +
> > +- When a new pod is brought up, its QoS class is firstly determined.
> > +- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
>
> will this happen before or after image pulling?
>
> /cc @vishh <https://github.com/vishh> - I would like to be able to in the
> future charge image pulling to the pod cgroup.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26751/files/86f33723eebe28602e420e4118a1a13e986b6c79#r68434206>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKCpbqYXhEBU8FuM31nc8RF73t0Lmks5qPBfKgaJpZM4ItEyP>
> .
>

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
+- This interface will be pluggable. Kubelet will support both systemd and raw cgroups based __cgroup__ drivers. We will be using the --cgroup-driver flag proposed in the [Systemd Node Spec](kubelet-systemd.md) to specify the cgroup driver.
+- We inject creation and deletion of pod level cgroups into the pod workers.
+- As new pods are added QoS class cgroup parameters are updated to match the resource requests by the Pods.
```
s/Pods/pods

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+#### Container level cgroups
+
+Have docker manager create container cgroups under pod level cgroups. For determining the parent pod cgroup name and path we have a separate implementation for systemd and raw cgroup systems. This can be easily implemented in raw cgroup systems as now docker has support for specifying parent flag of cgroups. In systemd systems we would have to add support for specifying the parent slice during sc ...
```
I would phrase this by stating that with the docker runtime, we will pass `--cgroup-parent` using the syntax expected for the corresponding `cgroup-driver` the runtime was configured to use.  We need to be able to introspect that.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+#### Add Pod level metrics to Kubelet's metrics provider
+
+Update Kubelet’s metrics provider to include Pod level metrics. Use cAdvisor's cgroup subsystem information to determine various Pod level usage metrics.
```
Will we update the `kubelet` summary stats provider to return pod level usage as well.  Right now, its just containers.

derekwaynecarr replied:
(this would let me simplify some code in eviction)

dubstack replied:
I believe thats our plan. 
cc @vishh

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+
+## Rollout Plan
+
+This feature will be opt-in in v1.4 and an opt-out in v1.5. We recommend users to drain their nodes and opt-in, before switching to v1.5, which will result in a no-op when v1.5 kubelet is rolled out.
```
To clarify, opt-in does not mean alpha?

dubstack replied:
cc @vishh

derekwaynecarr commented:
@dubstack - this looks pretty good, just a few questions, comments, typos, and clarifications ;-)

/cc @kubernetes/rh-cluster-infra @mrunalp for any other comments.

davidopp commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+- The Best Effort and Burstable pods are strictly restricted from bursting and hogging resources and thus `G` Pods are guaranteed resource isolation.
+- Best Effort pods are treated as lowest priority. So for the `BE` QoS cgroup we set cpu shares to the lowest possible value ie.2. This ensures that the best effort containers get a relatively small share of cpu time.
+- Also we don't set any quota on the cpu resources as the containers on the `BE` pods can use any amount of free resources on the node.
+- Having memory limit of `BE` cgroup as (Allocatable - summation of memory limits of Guaranteed and Burstable pods) would result in `BE` pods becoming more susceptible to being OOM killed. As more `G` and `Bu` pods are scheduled kubelet will more likely kill `BE` pods, even if the `G` and `Bu` pods are using less than their request since we will be dynamically reducing the size of `BE` m.limit_in ...
```
I was confused about the way you set memory limit of BE (and to some extent for Bu, but let's just focus on BE). Let's say you have a 2G machine running a 1.9G-limit G pod, but it uses only 0.5G, and you schedule a BE pod on this node. IIUC, the current proposal would kill the BE pod once it starts using 0.1G. If that's correct, this seems like it wastes utilization, as 1.4G of the machine is unused. Wouldn't it be better to let the BE pod use as much memory as it wants, and then if there is an OOM you obviously kill it instead of the G pod? It seems that unlike today, to get high utilization with this scheme you would need some kind of "resource estimator" that reduces the limit of G pods that are not using what they asked for--otherwise resources get wasted.

davidopp replied:
And if the goal is to avoid at all costs the system OOM killer being invoked, it seems like you could still have something like @derekwaynecarr 's evictor evict when the machine gets close to OOMing based on actual  memory usage.

dubstack replied:
Yes you are right with the example. The current proposal prioritizes resource guarantees to Guaranteed pods over resource utilization. We want that Guaranteed pods should always have the amount of resources they asked for. @dchen1107 @vishh and @derekwaynecarr would be able to provide more arguments in support of the current decision.

davidopp replied:
But I think Guatanteed pods are guaranteed access to their request/limit either way. As long as Guaranteed pods are the last to be killed (by system OOM handler or user-space pre-OOM handler) they should be definition never be killed since limit==request and the scheduler will never put a new Guaranteed pod on a machine if machine's capacity minus sum of requests of the pods already on the machine is less than the new Guaranteed pod's request. (I'm ignoring memory usage by things outside the scope of the pod accounting system described here, such as docker daemon.) 

To put it another way, can you describe a scenario (e.g. configuration of pods running on the machine) where having no memory limit on the Bu and BE cgroups would lead to Guaranteed pods receiving a weaker resource guarantee relative to the approach proposed here?

dubstack commented:
@derekwaynecarr Thanks a lot for all your comments. I have addressed all but a couple of comments. PTAL.

derekwaynecarr commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,396 @@
+- The `BE` and `Bu` pods are strictly restricted from bursting and hogging resources and thus `G` Pods are guaranteed resource isolation.
+- `BE` pods are treated as lowest priority. So for the `BE` QoS cgroup we set cpu shares to the lowest possible value ie.2. This ensures that the `BE` containers get a relatively small share of cpu time.
+- Also we don't set any quota on the cpu resources as the containers on the `BE` pods can use any amount of free resources on the node.
+- Having memory limit of `BE` cgroup as (Allocatable - summation of memory limits of `G` and `Bu` pods) would result in `BE` pods becoming more susceptible to being OOM killed. As more `G` and `Bu` pods are scheduled kubelet will more likely kill `BE` pods, even if the `G` and `Bu` pods are using less than their request since we will be dynamically reducing the size of `BE` m.limit_in_bytes. But  ...
```
This should be summation of memory requests of `G` and `Bu` pods.  We only guarantee requests.

I am also sympathetic to @davidopp argument around resource utilization.

dubstack replied:
Ack! Thanks!

euank commented on docs/proposals/pod-resource-management.md:
```diff
@@ -0,0 +1,394 @@
+/Pod1/cpu.shares = 110m  
+/Pod2/cpu.quota = 20m  
+/Pod2/cpu.shares = 20m  
+/Pod1/m.limit = 3Gi  
```
Is this shorthand for `memory.limit_in_bytes`?

dubstack replied:
Yes

euank replied:
I can understand not wanting `limit_in_bytes`, but I'd prefer `memory` be written in full.

dubstack replied:
Sounds Good

davidopp commented:
I'd like to ask that my comment about utilization be addressed before this is merged. It seems that you're sacrificing utilization without getting any more protection in return. (Of course, I may be missing some subtlety that explains it.) Copying my comment here since it got buried by Github when the PR was updated:

> I was confused about the way you set memory limit of BE (and to some extent for Bu, but let's just focus on BE). Let's say you have a 2G machine running a 1.9G-limit G pod, but it uses only 0.5G, and you schedule a BE pod on this node. IIUC, the current proposal would kill the BE pod once it starts using 0.1G. If that's correct, this seems like it wastes utilization, as 1.4G of the machine is unused. Wouldn't it be better to let the BE pod use as much memory as it wants, and then if there is an OOM you obviously kill it instead of the G pod? It seems that unlike today, to get high utilization with this scheme you would need some kind of "resource estimator" that reduces the limit of G pods that are not using what they asked for--otherwise resources get wasted.

> And if the goal is to avoid at all costs the system OOM killer being invoked, it seems like you could still have something like @derekwaynecarr 's evictor evict when the machine gets close to OOMing based on actual memory usage.

> As long as Guaranteed pods are the last to be killed (by system OOM handler or user-space pre-OOM handler) they should be definition never be killed since limit==request and the scheduler will never put a new Guaranteed pod on a machine if machine's capacity minus sum of requests of the pods already on the machine is less than the new Guaranteed pod's request. (I'm ignoring memory usage by things outside the scope of the pod accounting system described here, such as docker daemon.)

> To put it another way, can you describe a scenario (e.g. configuration of pods running on the machine) where having no memory limit on the Bu and BE cgroups would lead to Guaranteed pods receiving a weaker resource guarantee relative to the approach proposed here?

dubstack commented:
Yes I realize that your points are really valid. But I am not exactly sure about the arguments on the basis of which the current policy decisions were made. @dchen1107 @vishh and @derekwaynecarr would be able to provide any subtleties which lead to the current decision. And we would definitely address your concerns before this gets merged in.

davidopp commented:
Just realized I never followed up on this...
@dubstack @dchen1107 @vishh and I discussed this a few weeks ago and decided to make the thing I was talking about be configurable (per cluster) rather than hard-coded. So you can decide whether you value utilization more or very predictable tail latency for latency-sensitive servers. I think we decided the default would be higher utilization.

dubstack commented:
Yes I would be adding a section highlighting the change. Something along the lines of this: 

```We add a flag `qos-memory-overcommitment` to kubelet which would allow users to configure the percentage of memory overcommitment on the node. We have the default as 100, so by default we allow complete overcommitment on the node and let the BE pod use as much memory as it wants, and not reserve any resources for the G and Bu pods. As expected if there is an OOM in such a case we first kill the BE pods before the G and Bu pods. 
On the other hand if a user wants to ensure very predictable tail latency for latency-sensitive servers he would need to set qos-memory-overcommitment to a really low value(preferrably 0). In this case memory resources would be reserved for the G and Bu pods and BE pods would be able to use only the left over memory resource.```

vishh commented:
@davidopp PTAL at this patch - https://github.com/kubernetes/kubernetes/pull/26751/commits/ad8e2c2d16bb3f2d4b5123729d60ce7064390df4

@derekwaynecarr This patch is ready to go, unless you have any more comments.

dubstack commented:
@davidopp @derekwaynecarr ping.

davidopp commented:
Don't wait on me -- based on https://github.com/kubernetes/kubernetes/pull/26751#issuecomment-231557035 it sounds like you've addressed my concern, and I won't have time to re-read the doc soon.

dubstack commented:
Reattaching LGTM after running hack/update-munge-docs.sh to modify the image warning urls at the top.
